compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs

```diff
@@ -790,36 +928,151 @@ impl Context {
         } else {
             // Initialize the new array with the values from the old array
             result_block_id = self.block_id(result_id);
-            let init_values = try_vecmap(0..len, |i| {
+            let init_values = try_vecmap(0..array_len, |i| {
                 let index = AcirValue::Var(
                     self.acir_context.add_constant(FieldElement::from(i as u128)),
                     AcirType::NumericType(NumericType::NativeField),
                 );
                 let var = index.into_var()?;
                 let read = self.acir_context.read_from_memory(block_id, &var)?;
-                Ok(AcirValue::Var(read, AcirType::NumericType(NumericType::NativeField)))
+                Ok::<AcirValue, RuntimeError>(AcirValue::Var(
+                    read,
+                    AcirType::NumericType(NumericType::NativeField),
+                ))
             })?;
-            self.initialize_array(result_block_id, len, Some(&init_values))?;
+            self.initialize_array(
+                result_block_id,
+                array_len,
+                Some(AcirValue::Array(init_values.into())),
+            )?;
         }
 
-        // Write the new value into the new array at the specified index
-        self.acir_context.write_to_memory(result_block_id, &var_index, &store_value)?;
+        // TODO: Need to add support for dynamic indices with non-homogenous slices
```

Comment:
I will make an issue for this. I chose not to do it as a part of this PR as it is not totally obvious as we have both a capacity based on the witness and then a dynamically tracked length.

Reply:
https://github.com/noir-lang/noir/issues/2752